### PR TITLE
disableUserAccount Endpoint added

### PR DESCRIPTION
**Pull request by daisyzarzo**
relates to 
https://github.com/bespoken/source-api/issues/188

### CODE REVIEW

File: src/main/firebase/FirebaseController.ts [hunk 1]
```diff
@@ -155,6 +170,65 @@ export class FirebaseDatabase {
...
+
+    async  disableSourcesMonitoringAnValidationAndDisableUserStatus(userId: string, sources: any) {
+        let sourcePromises: any = [];
+        let sourcesToDisablePromises: any = [];
+        let disabledSourcesId: string[] = [];
+        try {
+            if (sources) {
+                for (const key in sources) {
+                    if (sources.hasOwnProperty(key) && sources[key] === "owner") {
+                        sourcePromises.push(this.getSource({ id: key, secretKey: "" }));
```
Comment (jperata) [inline src/main/firebase/FirebaseController.ts]: in this query, are you searching for sources that have a key but secretKey is empty?, there shouldn't be any source that fit's that condition

File: src/main/firebase/FirebaseController.ts [hunk 2]
```diff
@@ -155,6 +170,65 @@ export class FirebaseDatabase {
...
+                    }
+                }
+                const sourcesResutlt: any = await Promise.all(sourcePromises);
```
Comment (jperata) [inline src/main/firebase/FirebaseController.ts]: const sourcesResult

File: src/main/services/v1/VirtualDevice.ts [hunk 3]
```diff
@@ -34,7 +42,17 @@ export const postUpdateVirtualDeviceStatus = (token: string, status: string) =>
     });
 }
 
-
+export const updateVirtualDevicesStatus = async (virtualDevices: any, status: virtualDeviceStatus) => {
+    let virtualDevicePromises: any = [];
```
Comment (jperata) [inline src/main/services/v1/VirtualDevice.ts]: const

File: src/main/services/v1/VirtualDevice.ts [hunk 4]
```diff
@@ -34,7 +42,17 @@ export const postUpdateVirtualDeviceStatus = (token: string, status: string) =>
...
+    for (const key in virtualDevices) {
+        if (virtualDevices.hasOwnProperty(key)) {
+            const element = virtualDevices[key];
+            virtualDevicePromises.push(element.token, status);
+        }
+    }
+    await Promise.all(virtualDevicePromises);
+    return Promise.resolve();
```
Comment (jperata) [inline src/main/services/v1/VirtualDevice.ts]: when you use await/async, you don't need to return Promise.resolve()

File: src/test/firebase/FirebaseController.test.ts [hunk 5]
```diff
@@ -6,6 +6,8 @@ import * as MockFirebase from "./MockFirebase";
 import * as Source from "../../main/models/Source";
 import { Message } from "../../main/models/Message";
 import * as FirebaseController from "../../main/firebase/FirebaseController";
+// import { updateVirtualDevicesStatus } from "../../main/services/v1/VirtualDevice";
```
Comment (jperata) [inline src/test/firebase/FirebaseController.test.ts]: remove

File: src/main/firebase/FirebaseController.ts [hunk 6]
```diff
@@ -155,6 +170,65 @@ export class FirebaseDatabase {
...
+                        disabledSourcesId.push(source.id);
+                    }
+                }
+            }
+            sourcesToDisablePromises.push(
+                this.updateUser(userId, { disabled_sources: disabledSourcesId, status: userStatus.Disabled })
+            );
+
+            await Promise.all(sourcesToDisablePromises);
+            return Promise.resolve();
```
Comment (jperata) [inline src/main/firebase/FirebaseController.ts]: this return is not necessary

File: src/main/firebase/FirebaseController.ts [hunk 7]
```diff
@@ -155,6 +170,65 @@ export class FirebaseDatabase {
...
+        } catch (error) {
+            console.error("There was a problem trying to deactivate source monitoring or update user status, userID : ", userId);
+            return Promise.reject(error);
```
Comment (jperata) [inline src/main/firebase/FirebaseController.ts]: throw error instead of return Promise.reject

File: src/main/firebase/FirebaseController.ts [hunk 8]
```diff
@@ -155,6 +170,65 @@ export class FirebaseDatabase {
...
+            }
+        } catch (error) {
+            console.error("error for user " + userId);
+            console.error(error);
+            return Promise.reject(error);
+        }
+        return Promise.resolve();
+    }
+
+    async  disableSourcesMonitoringAnValidationAndDisableUserStatus(userId: string, sources: any) {
```
Comment (jperata) [inline src/main/firebase/FirebaseController.ts]: according to your function name, you disable the user status here, but in your code, that update is outside the function.
Change the name accordingly, but +1 to have explicit function names

Small fix on ( disableSourcesMonitoringAndValidation)

File: src/main/firebase/FirebaseController.ts [hunk 9]
```diff
@@ -155,6 +170,65 @@ export class FirebaseDatabase {
...
+            if (user.virtualDevices) {
+                console.log(user.virtualDevices);
+
+                await updateVirtualDevicesStatus(user.virtualDevices, virtualDeviceStatus.Suspended);
+                console.log("VirtualDevices state  updated to Suspended for the user: " + userId);
+            }
+        } catch (error) {
+            console.error("error for user " + userId);
+            console.error(error);
+            return Promise.reject(error);
```
Comment (jperata) [inline src/main/firebase/FirebaseController.ts]: throw error

File: src/main/firebase/FirebaseController.ts [hunk 10]
```diff
@@ -155,6 +170,65 @@ export class FirebaseDatabase {
...
+        }
+        return Promise.resolve();
```
Comment (jperata) [inline src/main/firebase/FirebaseController.ts]: not necessary

File: src/main/firebase/FirebaseController.ts [hunk 11]
```diff
@@ -109,6 +116,14 @@ export class FirebaseDatabase {
         }
     }
 
+    async updateSource(id: string, props: any): Promise<any> {
+        try {
+            return this.db.ref().child("/sources/" + id).update(props)
```
Comment (jperata) [inline src/main/firebase/FirebaseController.ts]: in order for the try to work you need an await here

File: src/main/firebase/FirebaseController.ts [hunk 12]
```diff
@@ -109,6 +116,14 @@ export class FirebaseDatabase {
...
+        } catch (error) {
+            return Promise.reject(error);
```
Comment (jperata) [inline src/main/firebase/FirebaseController.ts]: if you are only using the catch to throw the error again, you can just remove the try/catch

File: src/main/firebase/FirebaseController.ts [hunk 13]
```diff
@@ -204,9 +202,9 @@ export class FirebaseDatabase {
                 }
                 const sourcesResutlt: any = await Promise.all(sourcePromises);
                 for (const result of sourcesResutlt) {
-                    const source = result.val();
-
-                    if (source.validation_enabled) {
+                    const source = result.result;
+                    console.log(source);
```
Comment (jperata) [inline src/main/firebase/FirebaseController.ts]: remove

File: src/main/firebase/FirebaseController.ts (hunk 1 above)
Comment (daisyzarzo) [reply src/main/firebase/FirebaseController.ts]: image
the reason is that the funcion this.getSource need a obj:SourceObj,
but only use the id property

**Pull request merged**